crates/globwatch/src/lib.rs

```
@@ -0,0 +1,358 @@
+                                    None
+                                } else {
+                                    event!(parent: &span, Level::TRACE, "yielding {:?}", e);
+                                    Some(e)
+                                }
+                            }
+                            Either::Right(WatcherCommand::Flush(tx)) => {
+                                // create file in flush dir
+                                let flush_id =
+                                    flush_id.fetch_add(1, std::sync::atomic::Ordering::SeqCst);
+                                let flush_file = flush_dir.join(format!("{}", flush_id));
+                                if let Err(e) = std::fs::File::create(flush_file) {
+                                    warn!("failed to create flush file: {}", e);
+                                } else {
+                                    flush.lock().expect("no panic").insert(flush_id, tx);
+                                }
+                                None
+                            }
+                            Either::Right(WatcherCommand::Watcher(change)) => {
+                                event!(parent: &span, Level::TRACE, "change {:?}", change);
+                                Self::handle_config_change(
+                                    &mut watcher.lock().expect("no panic"),
+                                    change,
+                                );
+                                None
+                            }
+                        }
+                    }
+                })
+                .timeout_at(token),
+        )
+    }
+
+    #[tracing::instrument(skip(watcher))]
+    fn handle_config_change(watcher: &mut T, config: WatcherChange) {
+        match config {
+            WatcherChange::Include(glob, _) => {
+                for p in glob_to_paths(&glob) {
+                    if let Err(e) = watcher.watch(&p, notify::RecursiveMode::Recursive) {
```

What happens if we fail to watch something? What is the fallback if we later ask "did any of these files change"?

---

It depends per-watcher. For FsEvents on macOS.

Although you're right we will need to notify the caller about failed watches..

---

@gsoltis I am currently returning a grpc error if watching failed but perhaps the correct option is to just silently return an 'ok' here?

---

Returning an error if watching failed (`NotifyOutputsWritten`) is fine. The important part is when we later ask if anything has changed (`GetChangedOutputs`). If any errors have occurred, including failing to start watching, we _must_ return that everything changed, because we don't know any better.

---

I believe this has been handled properly now, by capturing globs with failed watch notices and immediately removing them again from the hashmaps.